persist/fs/write.go

```diff
@@ -195,33 +235,141 @@ func (w *writer) writeAll(
+	// NB(r): Write the index file in order, in the future we could write
+	// these in order to avoid this sort at the end however that does require
+	// significant changes in the storage/databaseShard to store things in order
+	// which would sacrifice O(1) insertion of new series we currently have.
+	//
+	// Probably do want to do this at the end still however so we don't stripe
+	// writes to two different files during the write loop.
+	sort.Sort(indexEntries(w.indexEntries))
+
+	summariesApprox := float64(len(w.indexEntries)) * w.summariesPercent
+	summaryEvery := 0
+	if summariesApprox > 0 {
+		summaryEvery = int(math.Floor(float64(len(w.indexEntries)) / summariesApprox))
+	}
+
+	// Write the index entries and calculate the bloom filter
+	n, p := uint(w.currIdx), w.bloomFilterFalsePositivePercent
+	m, k := xsets.BloomFilterEstimate(n, p)
+	bloomFilter := xsets.NewBloomFilter(m, k)
+
+	var offset int64
+	for i := range w.indexEntries {
+		id := w.indexEntries[i].id.Data().Get()
+
+		entry := schema.IndexEntry{
+			Index:    w.indexEntries[i].index,
+			ID:       id,
+			Size:     w.indexEntries[i].size,
+			Offset:   w.indexEntries[i].offset,
+			Checksum: w.indexEntries[i].checksum,
+		}
+
+		w.encoder.Reset()
+		if err := w.encoder.EncodeIndexEntry(entry); err != nil {
+			return err
+		}
+
+		data := w.encoder.Bytes()
+		if _, err := w.indexFdWithDigest.Write(data); err != nil {
+			return err
+		}
+
+		// Add to the bloom filter, note this must be zero alloc or else this will
+		// cause heavy GC churn as we flush millions of series at end of each
+		// time window
+		bloomFilter.Add(id)
+
+		if i%summaryEvery == 0 {
+			// Capture the offset for when we write this summary back, only capture
+			// for every summary we'll actually write to avoid a few memcopies
+			w.indexEntries[i].offsetForSummary = offset
+		}
+
+		offset += int64(len(data))
+	}
+
+	// Write summaries and start zeroing out memory to avoid holding onto refs
+	summaries := 0
+	for i := range w.indexEntries {
+		if i%summaryEvery != 0 {
+			w.indexEntries[i].releaseRefs()
+			continue
+		}
+
+		summary := schema.IndexSummary{
+			Index:            w.indexEntries[i].index,
+			ID:               w.indexEntries[i].id.Data().Get(),
+			IndexEntryOffset: w.indexEntries[i].offsetForSummary,
+		}
+
+		w.encoder.Reset()
+		if err := w.encoder.EncodeIndexSummary(summary); err != nil {
+			return err
+		}
+
+		data := w.encoder.Bytes()
+		if _, err := w.summariesFdWithDigest.Write(data); err != nil {
+			return err
+		}
+
+		w.indexEntries[i].releaseRefs()
+		summaries++
+	}
+
+	// Reset summaries slice to avoid allocs for next shard flush, this avoids
+	// leaking memory as we've released all refs while writing summaries
+	w.indexEntries = w.indexEntries[:0]
+
+	// Write the bloom filter bitset out
+	err := bloomFilter.BitSet().Write(w.bloomFilterFdWithDigest)
+	if err != nil {
+		return err
+	}
+
 	info := schema.IndexInfo{
```

Comment:
Should we have constructors for these msgpack structs? Seems like it would be easy to leave out a field by accident and corrupt the whole thing

Reply:
There's usually only a single place where we ever construct these and the unit tests are pretty rigorous.  Since there's only one piece of code constructing these msgpack structs I'd prefer the readability when setting the fields rather than the safety added of a constructor (where you don't know what each arg is when reading the code necessarily and maybe you specify the wrong local variable to the wrong argument).

If tons of different places constructed these types then I would agree with you.